skimage/morphology/_rolling_ball_cy.pyx

```diff
@@ -0,0 +1,34 @@
+import numpy as np
+cimport numpy as np
+
+cimport cython
+
+ctypedef np.float_t DTYPE
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def apply_kernel(np.ndarray[DTYPE, ndim=4] windows,
+                 np.ndarray[DTYPE, ndim=2] kernel,
+                 np.ndarray[DTYPE, ndim=2] sagitta):
+    cdef np.ndarray out_array = np.zeros((windows.shape[0], windows.shape[1]))
+    cdef DTYPE[:, :] out_data = out_array
+    cdef int kernel_size = kernel.size
+    cdef np.ndarray[DTYPE, ndim=1] tmp_array = np.empty(kernel_size)
+    cdef DTYPE[:] tmp = tmp_array
+    cdef int idx, tmp_idx, im_x, im_y, kern_x, kern_y
+    cdef DTYPE min_value
+
+    for im_y in range(windows.shape[0]):
+        for im_x in range(windows.shape[1]):
+            for kern_y in range(kernel.shape[0]):
+                for kern_x in range(kernel.shape[1]):
+                    tmp_idx = kern_y * kernel.shape[1] + kern_x
+                    tmp[tmp_idx] = (windows[im_y, im_x, kern_y, kern_x] + sagitta[kern_y, kern_x]) * kernel[kern_y, kern_x]
+            
+            min_value = tmp[0]
+            for idx in range(kernel_size):
+                if min_value > tmp[idx]:
+                    min_value = tmp[idx]
+            out_data[im_y, im_x] = min_value
```

Comment:
You're searching for the global minimum inside `tmp` here, correct? Why not skip the intermediate step with the buffer and directly compare what you're assigning to `tmp[tmp_idx]` with the previous `min_value`. You'd need a proper way to initialize `min_value` though. In case you stay with float, `-np.inf` should work otherwise use the global minimum of the image?

Reply:
Indeed. This makes the code A LOT faster; looks like the compiler has a really easy time optimizing this.

I played around with some optimization of this. I tried adding BLAS, and skipping elements of the (square) kernel that are not inside the (contained) ellipsis. Both versions were significantly slower in comparison.